doc/module-system: Introduce concepts (#240531)
NixOS/nixpkgs: doc/module-system/module-system.chapter.md

# Module System {#module-system}

::: {.note}
This chapter is new and not complete yet. For a gentle introduction to the module system, in the context of NixOS, see [Writing NixOS Modules](https://nixos.org/manual/nixos/unstable/index.html#sec-writing-modules) in the NixOS manual.
:::

## Introduction {#module-system-introduction}

The module system is a language for handling configuration, implemented as a Nix library.

It was historically known as the NixOS module system.

Its purpose is to "extend" the Nix language with

- **dynamic type checking** to find mistakes early on, improving error messages;

> - **documentation**, so that each option has a description, conveniently located in the file that declares it

Reviewer: It’s the beginning of the introduction, it’s probably fine to just call it a “configuration option” since we’ve already mentioned that modules are for configuration.

- **modularity** or **aspect oriented programming**: separate files can all contribute to the same option.

The following sections are dense introductory reference material.

<!-- TODO: link the tutorial
   If you're learning to work with the module system, consider following the introductory tutorial to get a feel for the module system. -->

## Concepts {#module-system-concepts}

The following sections briefly define the concepts that make up the module system.

<!-- keep them short -->

### Module Evaluation {#module-system-concept-evaluation}

<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->

> Module evaluation is the process of turning a set of modules into option values.

Reviewer: I think it’s separate because the options may not have values until they’re evaluated so what’s being discussed is the process of reifying them.

> Module evaluation is the process of turning a set of modules into option values.

Reviewer 2: @roberth I'd add a definition of option values here, like `option values, ...`

> The module system can be seen as an embedded domain specific language (eDSL) for configuration. The defining feature of an eDSL is that they are evaluated by a host language, which in this case is the Nix expression language. So _module evaluation_ is implemented using Nix evaluation, by means of the [`evalModules`] function.

Author: Not sure if the introduction benefits from defining eDSL. That seems too detailed to me.

I agree that the short definition of evaluation is suitable for a glossary, but that's not actionable now (and doesn't imply that it should be moved away from here either).

### Configuration {#module-system-concept-configuration}

<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->

A configuration is the result of [module evaluation](#module-system-concept-evaluation).

It consists of a set of options. Each option is identified by an [attribute path], and may produce a single value (which may in turn be a composite value such as an attribute set).

The module system provides its caller with _all_ the options and their values. It is up to the [application] to assign meaning to the options and to pick options (typically one option) that represents the value of the entire configuration.

### Module {#module-system-concept-module}

<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->

> A module is a function that produces part of the configuration.

Author: > goes against the intuition

That intuition is not a good starting point, and I don't think reference documentation is a good place to dispel all "wrong" intuitions.
I'd only do that if newcomers come with wrong preconceptions, such as "`=` is assignment". This is nowhere near that level.

Perhaps this sentence is more useful when referencing the module arguments and their use.
That's getting a bit ahead of the more mechanical explanation that follows, but maybe that's what the reader needs.

```suggestion
A module is a function that produces part of the configuration. The module arguments give the module access to the final configuration, to allow it to perform configuration logic.
```

> A module is a function that produces part of the configuration.

Reviewer 2: >  don't think reference documentation is a good place to dispel all "wrong" intuitions.

Btw this is not reference documentation right? See also my comment about moving this to nix.dev under "Concepts"

> A module is a function that produces part of the configuration.

Reviewer 2: See https://diataxis.fr/reference-explanation/#work-and-study

> A module can be written using multiple, similar [syntaxes](#module-syntax), which are converted internally to the most general syntax, which is a function that returns an attribute set with specific attribute names.

Author: I would consider the _representation_ of a module to be the most general form, as in "all modules are representable as functions that return `{ config, options, meta }` etc".

Here's an alternative way of saying this:

```suggestion
The most general representation of a module is a function that returns an attribute set containing specific attributes such as `options` and `config`, but as a user of the module system, you may write modules using more convenient [syntaxes](#module-syntax).
```

> A module can be written using multiple, similar [syntaxes](#module-syntax), which are converted internally to the most general syntax, which is a function that returns an attribute set with specific attribute names.

Reviewer 2: What are these syntaxes?

> These attributes, the [module attributes] are then processed in order to produce the [module arguments], which are passed to the modules.

Reviewer: ```suggestion
These attributes, the [module attributes], are then processed in order to produce the [module arguments], which are passed to the modules.
```

As you can see, module system evaluation is a cyclic process. This works thanks to Nix's lazy evaluation.

### Data flow {#module-system-data-flow}

<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->

The following diagram shows the data flows that occur within a module system evaluation.

Multiplicity is not modeled. For example `.options -> options` is a many-to-many transformation where any option declaration can affect any option.

```mermaid

> flowchart TD

Reviewer 3: The nodes could have links to anchors in the text to provide more details on what each one is for and what it means.

allModules["all modules"]
dotimports[".imports"]
dotoptions[".options"]
dotconfig[".config"]

subgraph evalModules["evalModules arguments"]
modules
prefix
more["..."]
end

modules --> allModules

allModules -->|has| dotimports
allModules -->|has| dotoptions
allModules -->|has| dotconfig

dotimports -->|closure| allModules

dotoptions -->|merge| options

options -->|default| definitions
options -->|type.merge| config

dotconfig --> definitions

definitions -->|merge| config

options -->|argument| allModules
config -->|argument| allModules

subgraph configuration["returned configuration"]
retOptions["options"]
retConfig["config"]
retModule["_module"]
retMore["..."]
end

options --> retOptions
config -->|remove ._module| retConfig
config -->|_module| retModule
```

### Module syntax {#module-syntax}

<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->

A module can be specified as an attribute set or a function.

> The attribute set may be in canonical or shorthand form.

Reviewer 2: Examples please :)
